File: TouchNothing/Services/LocalUserStore.swift
import Foundation

enum LocalUserStore {
    static var nickname: String? {
        get {
            guard let raw = UserDefaults.standard.string(forKey: StorageKeys.nickname) else {
                return nil
            }
            let normalized = NicknameValidator.normalized(raw)
            return normalized.isEmpty ? nil : normalized
        }
        set { UserDefaults.standard.set(newValue, forKey: StorageKeys.nickname) }
    }

    static var sessionToken: String? {
        get {
            guard let token = UserDefaults.standard.string(forKey: StorageKeys.sessionToken) else {
                return nil
            }
            return token.isEmpty ? nil : token
        }
        set { UserDefaults.standard.set(newValue, forKey: StorageKeys.sessionToken) }
    }

    static var isGuest: Bool {
        get { UserDefaults.standard.bool(forKey: StorageKeys.isGuest) }
        set { UserDefaults.standard.set(newValue, forKey: StorageKeys.isGuest) }
    }

    static var guestPin: String? {
        get {
            guard let pin = UserDefaults.standard.string(forKey: StorageKeys.guestPin) else {
                return nil
            }
            return PinInput.isValid(pin) ? pin : nil
        }
        set { UserDefaults.standard.set(newValue, forKey: StorageKeys.guestPin) }
    }

    static var rulesHidden: Bool {
        get { UserDefaults.standard.bool(forKey: StorageKeys.rulesHidden) }
        set { UserDefaults.standard.set(newValue, forKey: StorageKeys.rulesHidden) }
    }

    static var hasNickname: Bool {
        guard let nickname else { return false }
        return !nickname.isEmpty
    }

    static var hasActiveSession: Bool {
        hasNickname && sessionToken != nil
    }

    static func saveSession(nickname: String, sessionToken: String, isGuest: Bool = false, guestPin: String? = nil) {
        self.nickname = nickname
        self.sessionToken = sessionToken
        self.isGuest = isGuest
        self.guestPin = isGuest ? guestPin : nil
    }

    static func signOut() {
        nickname = nil
        sessionToken = nil
        isGuest = false
        guestPin = nil
        LeaderboardCache.clearAll()
    }
}

File: app/src/main/java/com/touchnothing/app/data/local/LocalUserStore.kt
package com.touchnothing.app.data.local

import android.content.Context
import com.touchnothing.app.util.NicknameValidator
import com.touchnothing.app.util.PinInput

class LocalUserStore(context: Context) {
    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private val leaderboardCache = LeaderboardCache(context)

    var nickname: String?
        get() {
            val raw = prefs.getString(StorageKeys.NICKNAME, null) ?: return null
            val normalized = NicknameValidator.normalized(raw)
            return normalized.ifEmpty { null }
        }
        set(value) {
            if (value == null) {
                prefs.edit().remove(StorageKeys.NICKNAME).apply()
                return
            }
            prefs.edit().putString(StorageKeys.NICKNAME, value).apply()
        }

    var sessionToken: String?
        get() {
            val token = prefs.getString(StorageKeys.SESSION_TOKEN, null) ?: return null
            return token.ifEmpty { null }
        }
        set(value) {
            if (value == null) {
                prefs.edit().remove(StorageKeys.SESSION_TOKEN).apply()
                return
            }
            prefs.edit().putString(StorageKeys.SESSION_TOKEN, value).apply()
        }

    var isGuest: Boolean
        get() = prefs.getBoolean(StorageKeys.IS_GUEST, false)
        set(value) = prefs.edit().putBoolean(StorageKeys.IS_GUEST, value).apply()

    var guestPin: String?
        get() {
            val pin = prefs.getString(StorageKeys.GUEST_PIN, null) ?: return null
            if (!PinInput.isValid(pin)) return null
            return pin
        }
        set(value) {
            if (value == null) {
                prefs.edit().remove(StorageKeys.GUEST_PIN).apply()
                return
            }
            prefs.edit().putString(StorageKeys.GUEST_PIN, value).apply()
        }

    var rulesHidden: Boolean
        get() = prefs.getBoolean(StorageKeys.RULES_HIDDEN, false)
        set(value) = prefs.edit().putBoolean(StorageKeys.RULES_HIDDEN, value).apply()

    val hasNickname: Boolean
        get() = !nickname.isNullOrEmpty()

    val hasActiveSession: Boolean
        get() = hasNickname && sessionToken != null

    fun saveSession(
        nickname: String,
        sessionToken: String,
        isGuest: Boolean = false,
        guestPin: String? = null,
    ) {
        this.nickname = nickname
        this.sessionToken = sessionToken
        this.isGuest = isGuest
        this.guestPin = if (isGuest) guestPin else null
    }

    fun signOut() {
        nickname = null
        sessionToken = null
        isGuest = false
        guestPin = null
        leaderboardCache.clearAll()
    }

    companion object {
        const val PREFS_NAME = "touch_nothing_prefs"
    }
}
